feat(learn): add word frequency challenge to rosetta section (#39424)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/10-coding-interview-prep/rosetta-code/word-frequency.english.md

</section>

## Instructions
<section id='instructions'>
Write a function to count the occurrences of each word and return the n most commons words along with the number of their occurences in decreasing frequency.
The function should return a 2D array with each of the elements in the following form: <code>[word, freq]</code>.  <code>word</code> should be the lowercase version of the word and <code>freq</code> the number denoting the count.
Your function should be case insensitive, for example, the strings "Hello" and "hello" should be treated the same. 
You can treat words that have special characters such as underscores, dashes, apostrophes, commas, etc., as distinct words. 
For example, given the string "Hello hello goodbye", your function should return <code>[['hello', 2], ['goodbye', 1]]</code>.
</section>

## Tests
<section id='tests'>

``` yml
tests:
  - text: <code>wordFrequency</code> should be a function.
    testString: assert(typeof wordFrequency == 'function');
  - text: <code>wordFrequency</code> should return an array.
    testString: assert(Array.isArray(wordFrequency("test")));

>   - text: <code>wordFrequency</code> should return the expected output for <code>("Hello hello world", 2)</code>

Reviewer: ```suggestion
  - text: <code>wordFrequency("Hello hello world", 2)</code> should return <code>[['hello', 2], ['world', 1]]</code>
```

    testString: assert.deepEqual(wordFrequency(example_1, 2), example_1_solution)

>   - text: <code>wordFrequency</code> should return the expected output for <code>("The quick brown fox jumped over the lazy dog", 4)</code>

Reviewer: ```suggestion
  - text: <code>wordFrequency("The quick brown fox jumped over the lazy dog", 4)</code> should return  <code>[['the', 2], ['brown', 1], ['dog', 1], ['fox', 1]]</code>
```

    testString: assert.deepEqual(wordFrequency(example_2, 4), example_2_solution)

>   - text: <code>wordFrequency</code> should return the expected output for <code>("Opensource opensource open-source open source", 1)</code>

Reviewer: ```suggestion
  - text: <code>wordFrequency("Opensource opensource open-source open source", 1)</code> should return 
 <code>[['opensource', 2]]</code>
```

    testString: assert.deepEqual(wordFrequency(example_3, 1), example_3_solution)

>   - text: <code>wordFrequency</code> should return the expected output for <code>("Apple application App apply applicative aPP aPPlE able", 3)</code>

Reviewer: ```suggestion
  - text: <code>wordFrequency("Apple application App apply applicative aPP aPPlE able", 3)</code> should return <code>[['app', 2], ['apple', 2], ['able', 1]]</code>
```

    testString: assert.deepEqual(wordFrequency(example_4, 3), example_4_solution)

>   - text: <code>wordFrequency</code> should return the expected output for <code>("c d a d c a b d d c", 4)</code>

Reviewer: ```suggestion
  - text: <code>wordFrequency("c d a d c a b d d c", 4)</code> should return <code>[['d', 4], ['c', 3], ['a', 2], ['b', 1]]</code>
```

    testString: assert.deepEqual(wordFrequency(example_5, 4), example_5_solution)

>   - text: <code>wordFrequency</code> should return an empty array for <code>("", 5)</code>

Reviewer: ```suggestion
  - text: <code>wordFrequency("", 5)</code> should return <code>[]</code>
```

---
id: 5e94a54cc7b022105bf0fd2c
title: Word frequency
challengeType: 5
isHidden: false
forumTopicId: 393913
---

## Description

<section id='description'>
Given a text string and an integer n, return the n most common words in the file (and the number of their occurrences) in decreasing frequency.
</section>

## Instructions

<section id='instructions'>
Write a function to count the occurrences of each word and return the n most commons words along with the number of their occurences in decreasing frequency.
The function should return a 2D array with each of the elements in the following form: <code>[word, freq]</code>.  <code>word</code> should be the lowercase version of the word and <code>freq</code> the number denoting the count.

> Your function should be case insensitive, for example, the strings "Hello" and "hello" should be treated the same. 

Reviewer 2: ```suggestion
The function should return an empty array, if no string is provided.
The function should be case insensitive, for example, the strings "Hello" and "hello" should be treated the same. 
```
